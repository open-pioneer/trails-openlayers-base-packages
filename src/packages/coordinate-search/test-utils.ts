// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { act, fireEvent } from "@testing-library/react";

export function showDropdown(projSelect: HTMLElement) {
    // open dropdown to include options in snapshot; react-select creates list of options in dom after opening selection
    act(() => {
        fireEvent.keyDown(projSelect, { key: "ArrowDown" });
    });
}

export function getCurrentOptions() {
    return Array.from(
        document.getElementsByClassName("coordinate-input-select__option")
    ) as HTMLElement[];
}

export function getCurrentOptionValues(options: HTMLElement[]) {
    const values: (string | null)[] = [];
    for (const opt of options) {
        values.push(opt.textContent);
    }
    return values;
}

export function getClearButton(coordinateSearchGroup: Element) {
    const buttonDiv = coordinateSearchGroup.querySelector(".chakra-input__right-element");
    if (!buttonDiv) {
        throw new Error("buttons not rendered");
    }
    const clearButton = buttonDiv.querySelector(".coordinate-input-clear-button");
    if (!clearButton) {
        throw new Error("clear button not rendered");
    }
    return clearButton;
}

export function getCopyButton(coordinateSearchGroup: Element) {
    const buttonDiv = coordinateSearchGroup.querySelector(".chakra-input__right-element");
    if (!buttonDiv) {
        throw new Error("buttons not rendered");
    }
    const copyButton = buttonDiv.querySelector(".coordinate-input-copy-button");
    if (!copyButton) {
        throw new Error("copy button not rendered");
    }
    return copyButton;
}
