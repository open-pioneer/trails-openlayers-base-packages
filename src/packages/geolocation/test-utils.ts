// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import OlMap from "ol/Map";
import { Circle, Fill, Stroke, Style } from "ol/style";
import { vi } from "vitest";
import { GeolocationController } from "./GeolocationController";

export function mockSuccessGeolocation(coords: number[]) {
    vi.spyOn(navigator, "geolocation", "get").mockReturnValue({
        clearWatch() {},
        getCurrentPosition() {},
        watchPosition(success) {
            setTimeout(() => {
                success({
                    coords: {
                        latitude: coords[0],
                        longitude: coords[1],
                        accuracy: 2500,
                        altitude: null,
                        altitudeAccuracy: null,
                        heading: null,
                        speed: null
                    }
                } as GeolocationPosition);
            }, 1);

            return 123;
        }
    } satisfies Partial<Geolocation>);
}

export function mockErrorGeolocation() {
    vi.spyOn(navigator, "geolocation", "get").mockReturnValue({
        clearWatch() {},
        getCurrentPosition() {},
        watchPosition(success, error) {
            setTimeout(() => {
                error?.({
                    code: 2,
                    message: "POSITION_UNAVAILABLE",
                    PERMISSION_DENIED: 1,
                    POSITION_UNAVAILABLE: 2,
                    TIMEOUT: 3
                });
            }, 1);

            return 123;
        }
    } satisfies Partial<Geolocation>);
}

export function setup() {
    const olMap = new OlMap();
    return {
        olMap,
        controller: new GeolocationController(olMap, (error) => {
            console.error("Unexpected error", error);
        })
    };
}

export function setupWithCustomProperties() {
    const olMap = new OlMap();
    const maxZoom: number = getCustomMaxZoom();
    const positionFeatureStyle: Style = getCustomPositionStyle();
    const accuracyFeatureStyle: Style = getCustomAccuracyStyle();
    const trackingOptions: PositionOptions = getCustomTrackingOptions();

    const controller = new GeolocationController(
        olMap,
        (error) => {
            console.error("Unexpected error", error);
        },
        trackingOptions
    );
    controller.setMaxZoom(maxZoom);
    controller.setPositionFeatureStyle(positionFeatureStyle);
    controller.setAccuracyFeatureStyle(accuracyFeatureStyle);
    return {
        olMap,
        controller
    };
}

export function getCustomMaxZoom() {
    return 20;
}
export function getCustomPositionStyle() {
    return new Style({
        image: new Circle({
            fill: new Fill({
                color: "rgba(255,255,255,0.4)"
            }),
            stroke: new Stroke({
                color: "#3399CC",
                width: 1.25
            }),
            radius: 5
        }),
        fill: new Fill({
            color: "rgba(255,255,255,0.4)"
        }),
        stroke: new Stroke({
            color: "#3399CC",
            width: 1.25
        })
    });
}

export function getCustomAccuracyStyle() {
    return new Style({
        image: new Circle({
            fill: new Fill({
                color: "rgba(255,255,255,0.4)"
            }),
            stroke: new Stroke({
                color: "#3399CC",
                width: 1.25
            }),
            radius: 5
        }),
        fill: new Fill({
            color: "rgba(255,255,255,0.4)"
        }),
        stroke: new Stroke({
            color: "#3399CC",
            width: 1.25
        })
    });
}

function getCustomTrackingOptions() {
    return {
        "enableHighAccuracy": true,
        "timeout": 20,
        "maximumAge": 50
    };
}
