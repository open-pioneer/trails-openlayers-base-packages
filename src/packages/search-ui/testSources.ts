// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { DataSource, Suggestion } from "./api";

interface Feature {
    bbox: number[];
    geometry?: { coordinates: [number, number]; type: string };
    properties: Readonly<Record<string, unknown>>;
    type: string;
    id: string;
}
const fakeStreetData = [
    {
        text: "Brückenweg 1"
    },
    {
        text: "Mauerstrasse 3"
    },
    {
        text: "Vogelgasse 4"
    },
    {
        text: "Kampweg 8"
    }
];
const fakeCityData = [
    {
        text: "Aachen"
    },
    {
        text: "Dortmund"
    },
    {
        text: "Langenfeld"
    },
    {
        text: "Düsseldorf"
    },
    {
        text: "Münster"
    }
];

const fakeRiverData = [
    {
        text: "Rhein"
    },
    {
        text: "Weser"
    },
    {
        text: "Ems"
    },
    {
        text: "Maas"
    },
    {
        text: "Elbe"
    }
];

const getFakeData = async (
    inputValue: string,
    responseData: { text: string }[],
    signal?: AbortSignal | undefined,
    duration = 500
) => {
    return new Promise<typeof responseData>((resolve, reject) => {
        setTimeout(() => {
            if (signal?.aborted) {
                reject(new DOMException("Aborted", "AbortError"));
            }
            const cp = responseData.filter((item) =>
                item.text.toLowerCase().includes(inputValue.toLocaleLowerCase())
            );
            resolve(cp);
        }, duration);
    });
};

const request = async (url: string, signal?: AbortSignal | undefined): Promise<Feature[] | []> => {
    try {
        const response = await fetch(url, { signal });
        if (!response.ok) {
            throw new Error("Request failed: " + response.status);
        }
        const result = await response.json();
        return result.features;
    } catch (error) {
        if (error == "AbortError") {
            console.debug("Previous searchquery has been canceled by the user.");
        } else {
            console.error(error);
        }
        return [];
    }
};
export class GeoSearchSource implements DataSource {
    label: string = "Bkg-Service";
    async search(
        inputValue: string,
        options?: { signal?: AbortSignal | undefined } | undefined
    ): Promise<Suggestion[]> {
        const signal = options?.signal;
        const url = this.#getUrl(inputValue);
        try {
            const features = await request(url, signal);
            return features.map((feature, idx) => ({
                id: idx,
                text: feature.properties?.text,
                properties: feature.properties
            })) as Suggestion[];
        } catch (error) {
            console.error(error);
            return [];
        }
    }
    #getUrl(inputValue: string): string {
        return `URL`;
    }
}

export class FakeStreetSource implements DataSource {
    label: string = "Streets";
    async search(
        inputValue: string,
        options?: { signal?: AbortSignal | undefined } | undefined
    ): Promise<Suggestion[]> {
        const result = await getFakeData(inputValue, fakeStreetData, options?.signal);
        const suggestions = result.map((item, idx) => ({
            id: idx,
            text: item.text
        }));
        return suggestions;
    }
}
export class FakeCitySource implements DataSource {
    label: string = "Cities";
    async search(
        inputValue: string,
        options?: { signal?: AbortSignal | undefined } | undefined
    ): Promise<Suggestion[]> {
        const result = await getFakeData(inputValue, fakeCityData, options?.signal);
        const suggestions = result.map((item, idx) => ({
            id: idx,
            text: item.text
        }));
        return suggestions;
    }
}

export class FakeRiverSource implements DataSource {
    label: string = "Rivers";
    async search(
        inputValue: string,
        options?: { signal?: AbortSignal | undefined } | undefined
    ): Promise<Suggestion[]> {
        const result = await getFakeData(inputValue, fakeRiverData, options?.signal);
        const suggestions = result.map((item, idx) => ({
            id: idx,
            text: item.text
        }));
        return suggestions;
    }
}
