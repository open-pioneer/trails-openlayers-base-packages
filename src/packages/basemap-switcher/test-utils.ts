// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { MapConfig, MapConfigProvider, MapRegistry } from "@open-pioneer/map";
import TileLayer from "ol/layer/Tile";
import OSM from "ol/source/OSM";
import Stamen from "ol/source/Stamen";
import { createService } from "@open-pioneer/test-utils/services";
import { MapRegistryImpl } from "@open-pioneer/map/MapRegistryImpl";
import { PackageContextProviderProps } from "@open-pioneer/test-utils/react";
import { screen, waitFor } from "@testing-library/react";

// used to avoid a "ResizeObserver is not defined" error
import ResizeObserver from "resize-observer-polyfill";
global.ResizeObserver = ResizeObserver;

export interface SimpleMapOptions {
    center?: { x: number; y: number };
    zoom?: number;
}

export const MAP_ID = "test";

export async function waitForMapMount(parentTestId = "base") {
    return await waitFor(async () => {
        const domElement = await screen.findByTestId(parentTestId);
        const container = domElement.querySelector(".ol-viewport");
        if (!container) {
            throw new Error("map not mounted");
        }
        return domElement;
    });
}
export async function setupMap(options?: SimpleMapOptions) {
    const mapId = "test";
    const mapConfig: MapConfig = {
        initialView: {
            kind: "position",
            center: options?.center ?? { x: 847541, y: 6793584 },
            zoom: options?.zoom ?? 10
        },
        projection: "EPSG:3857",
        layers: [
            {
                id: "b-1",
                title: "OSM",
                isBaseLayer: true,
                visible: false,
                layer: new TileLayer({
                    source: new OSM()
                })
            },
            {
                id: "b-2",
                title: "Toner",
                isBaseLayer: true,
                visible: true,
                layer: new TileLayer({
                    source: new Stamen({ layer: "toner" })
                })
            }
        ]
    };

    const registry = await createService(MapRegistryImpl, {
        references: {
            providers: [new MapConfigProviderImpl(mapId, mapConfig)]
        }
    });

    return { mapId, registry };
}

export function createPackageContextProviderProps(
    service: MapRegistry
): PackageContextProviderProps {
    return {
        services: {
            "map.MapRegistry": service
        }
    };
}

class MapConfigProviderImpl implements MapConfigProvider {
    mapId = "default";
    mapConfig: MapConfig;

    constructor(mapId: string, mapConfig?: MapConfig | undefined) {
        this.mapId = mapId;
        this.mapConfig = mapConfig ?? {};
    }

    getMapConfig(): Promise<MapConfig> {
        return Promise.resolve(this.mapConfig);
    }
}
