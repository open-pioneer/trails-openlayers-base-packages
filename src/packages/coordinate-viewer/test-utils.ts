// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { OlMapConfigurationProvider } from "@open-pioneer/experimental-ol-map";
import { OlMapRegistry } from "@open-pioneer/experimental-ol-map/services";
import { ServiceOptions } from "@open-pioneer/runtime";
import { PackageContextProviderProps } from "@open-pioneer/test-utils/react";
import { createService } from "@open-pioneer/test-utils/services";
import { waitFor, screen } from "@testing-library/react";
import { MapOptions } from "ol/Map";
import View from "ol/View";
import TileLayer from "ol/layer/Tile";
import OSM from "ol/source/OSM";

// used to avoid a "ResizeObserver is not defined" error
import ResizeObserver from "resize-observer-polyfill";
global.ResizeObserver = ResizeObserver;

export interface SimpleMapOptions {
    center?: [number, number];
    zoom?: number;
}

export const MAP_ID = "test";

export async function waitForMapMount(parentTestId = "base") {
    return await waitFor(async () => {
        const domElement = await screen.findByTestId(parentTestId);
        const container = domElement.querySelector(".ol-viewport");
        if (!container) {
            throw new Error("map not mounted");
        }
        return domElement;
    });
}

export async function setupMap(options?: SimpleMapOptions) {
    const mapOptions: MapOptions = {
        view: new View({
            projection: "EPSG:3857",
            center: options?.center ?? [847541, 6793584],
            zoom: options?.zoom ?? 10
        }),
        layers: [
            new TileLayer({
                source: new OSM(),
                properties: { title: "OSM" }
            })
        ]
    };

    const mapConfigProvider = await createService(MapConfigProvider, {
        properties: {
            mapOptions: mapOptions,
            mapId: MAP_ID
        }
    });
    const registry = await createService(OlMapRegistry, {
        references: {
            providers: [mapConfigProvider]
        }
    });

    return { mapId: MAP_ID, registry };
}

export function createPackageContextProviderProps(
    service: OlMapRegistry
): PackageContextProviderProps {
    return {
        services: {
            "ol-map.MapRegistry": service
        }
    };
}

class MapConfigProvider implements OlMapConfigurationProvider {
    mapId = "default";
    mapOptions: MapOptions = {};

    constructor(options: ServiceOptions) {
        if (options.properties.mapOptions) {
            this.mapOptions = options.properties.mapOptions as MapOptions;
        }
        if (options.properties.mapId) {
            this.mapId = options.properties.mapId as string;
        }
    }

    getMapOptions(): Promise<MapOptions> {
        return Promise.resolve(this.mapOptions);
    }
}
