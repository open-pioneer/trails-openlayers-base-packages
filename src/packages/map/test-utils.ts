// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { PackageContextProviderProps } from "@open-pioneer/test-utils/react";
import { createService } from "@open-pioneer/test-utils/services";
import { screen, waitFor } from "@testing-library/react";
import TileLayer from "ol/layer/Tile";
import OSM from "ol/source/OSM";
import {
    ExtentConfig,
    InitialViewConfig,
    LayerConfig,
    MapConfig,
    MapConfigProvider,
    MapRegistry,
    OlMapOptions
} from "./api";
import { MapRegistryImpl } from "./services";

// used to avoid a "ResizeObserver is not defined" error
import ResizeObserver from "resize-observer-polyfill";
global.ResizeObserver = ResizeObserver;

export interface SimpleMapOptions {
    center?: { x: number; y: number };
    zoom?: number;
    extent?: ExtentConfig;
    projection?: string;
    layers?: LayerConfig[];
    advanced?: OlMapOptions;

    noInitialView?: boolean;
    noProjection?: boolean;
}

export const MAP_ID = "test";

export async function waitForMapMount(parentTestId = "base") {
    return await waitFor(async () => {
        const domElement = await screen.findByTestId(parentTestId);
        const container = domElement.querySelector(".ol-viewport");
        if (!container) {
            throw new Error("map not mounted");
        }
        return domElement;
    });
}

export async function setupMap(options?: SimpleMapOptions) {
    const mapId = "test";
    const getInitialView = (): InitialViewConfig => {
        if (options?.extent) {
            return {
                kind: "extent",
                extent: options.extent
            };
        }
        return {
            kind: "position",
            center: options?.center ?? { x: 847541, y: 6793584 },
            zoom: options?.zoom ?? 10
        };
    };

    const mapConfig: MapConfig = {
        initialView: options?.noInitialView ? undefined : getInitialView(),
        projection: options?.noProjection ? undefined : options?.projection ?? "EPSG:3857",
        layers: options?.layers ?? [
            {
                title: "OSM",
                layer: new TileLayer({
                    source: new OSM()
                })
            }
        ],
        advanced: options?.advanced
    };

    const registry = await createService(MapRegistryImpl, {
        references: {
            providers: [new MapConfigProviderImpl(mapId, mapConfig)]
        }
    });

    return { mapId, registry };
}

export function createPackageContextProviderProps(
    service: MapRegistry
): PackageContextProviderProps {
    return {
        services: {
            "map.MapRegistry": service
        }
    };
}

class MapConfigProviderImpl implements MapConfigProvider {
    mapId = "default";
    mapConfig: MapConfig;

    constructor(mapId: string, mapConfig?: MapConfig | undefined) {
        this.mapId = mapId;
        this.mapConfig = mapConfig ?? {};
    }

    getMapConfig(): Promise<MapConfig> {
        return Promise.resolve(this.mapConfig);
    }
}
