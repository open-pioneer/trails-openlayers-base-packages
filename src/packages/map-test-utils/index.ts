// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { createService } from "@open-pioneer/test-utils/services";
import { screen, waitFor } from "@testing-library/react";
import TileLayer from "ol/layer/Tile";
import OSM from "ol/source/OSM";
import {
    ExtentConfig,
    InitialViewConfig,
    LayerConfig,
    MapConfig,
    MapConfigProvider,
    MapModel,
    MapRegistry,
    OlMapOptions
} from "@open-pioneer/map";
// Importing internals: needed for test support
import { MapRegistryImpl } from "@open-pioneer/map/services";

export interface SimpleMapOptions {
    center?: { x: number; y: number };
    zoom?: number;
    extent?: ExtentConfig;
    projection?: string;
    layers?: LayerConfig[];
    advanced?: OlMapOptions;

    noInitialView?: boolean;
    noProjection?: boolean;
}

/**
 * Waits until the open layers map has been mounted in the parent with the given id.
 */
export async function waitForMapMount(parentTestId = "base") {
    return await waitFor(async () => {
        const domElement = await screen.findByTestId(parentTestId);
        const container = domElement.querySelector(".ol-viewport");
        if (!container) {
            throw new Error("map not mounted");
        }
        return domElement;
    });
}

/**
 * Waits until the model has an initial extent.
 */
export async function waitForInitialExtent(model: MapModel) {
    if (model.initialExtent) {
        return;
    }

    await new Promise<void>((resolve, reject) => {
        model?.once("changed:initialExtent", () => {
            if (model?.initialExtent) {
                resolve();
            } else {
                reject(new Error("expected a valid extent"));
            }
        });
    });
}

/**
 * Creates a simple map registry service with exactly one map configuration.
 *
 * The map is configured by using the `options` parameter.
 *
 * Returns the map registry and the id of the configured map.
 */
export async function setupMap(options?: SimpleMapOptions) {
    // Always use "test" as mapId for unit tests
    const mapId = "test";

    const getInitialView = (): InitialViewConfig => {
        if (options?.extent) {
            return {
                kind: "extent",
                extent: options.extent
            };
        }
        return {
            kind: "position",
            center: options?.center ?? { x: 847541, y: 6793584 },
            zoom: options?.zoom ?? 10
        };
    };

    const mapConfig: MapConfig = {
        initialView: options?.noInitialView ? undefined : getInitialView(),
        projection: options?.noProjection ? undefined : options?.projection ?? "EPSG:3857",
        layers: options?.layers ?? [
            {
                title: "OSM",
                layer: new TileLayer({
                    source: new OSM()
                })
            }
        ],
        advanced: options?.advanced
    };

    const registry = await createService(MapRegistryImpl, {
        references: {
            providers: [new MapConfigProviderImpl(mapId, mapConfig)]
        }
    });

    return { mapId, registry };
}

/**
 * Creates (service name, service implementation)-pairs suitable for the `services`
 * option of the `PackageContextProvider`.
 *
 * This helper method can be used to avoid hard-coding service names used in the implementation.
 */
export function createServiceOptions(services: { registry: MapRegistry }): Record<string, unknown> {
    return {
        "map.MapRegistry": services.registry
    };
}

class MapConfigProviderImpl implements MapConfigProvider {
    mapId = "default";
    mapConfig: MapConfig;

    constructor(mapId: string, mapConfig?: MapConfig | undefined) {
        this.mapId = mapId;
        this.mapConfig = mapConfig ?? {};
    }

    getMapConfig(): Promise<MapConfig> {
        return Promise.resolve(this.mapConfig);
    }
}
