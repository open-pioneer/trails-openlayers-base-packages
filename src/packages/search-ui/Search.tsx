// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { Box, chakra, FormControl } from "@open-pioneer/chakra-integration";
import { MapModel, useMapModel } from "@open-pioneer/map";
import { CommonComponentProps, useCommonComponentProps } from "@open-pioneer/react-utils";
import { FC, useCallback, useEffect, useState } from "react";
import { ActionMeta, AsyncSelect, components, MenuProps, NoticeProps } from "chakra-react-select";
import { DataSource, Suggestion } from "./api";
import { SearchController } from "./SearchController";
import { HighlightOption } from "./HighlightOption";
import { useIntl } from "open-pioneer:react-hooks";

export interface SearchOption {
    value: string;
    label: string;
}
export interface SearchGroupOption {
    label: string;
    options: SearchOption[];
    priority?: number;
}

export interface SearchEvent {
    action: string;
    value?: SearchOption;
}

/**
 * This is for special properties of the Search component
 */
export interface SearchProps extends CommonComponentProps {
    /**
     * The id of the map.
     */
    mapId: string;
    sources: DataSource[];
    name?: string;
    placeholder?: string;
    closeMenuOnSelect?: boolean;
    searchTypingDelay?: number;
    showDropdownIndicator?: boolean;
    onSelect(event: SearchEvent): void;
    onClear(event: SearchEvent): void;
}

export const Search: FC<SearchProps> = (props) => {
    const {
        placeholder,
        closeMenuOnSelect,
        mapId,
        sources,
        searchTypingDelay,
        showDropdownIndicator,
        onSelect,
        onClear
    } = props;
    const { containerProps } = useCommonComponentProps("search", props);
    const { map } = useMapModel(mapId);
    const controller = useController(sources, map);
    const loadOptions = async (inputValue: string): Promise<SearchGroupOption[]> => {
        const suggestions = await controller!.search(inputValue);
        return mapSuggestions(suggestions, sources);
    };

    const debouncedLoadOptions = useCallback(
        debounce(async (inputValue: string, callback: (options: unknown) => void) => {
            try {
                const results = await loadOptions(inputValue);
                callback(results); // <-- Notice we added here the "await" keyword.
            } catch (e) {
                if (e instanceof Error) {
                    if (e.name == "AbortError") {
                        console.debug("Previous searchquery has been canceled by the user.");
                    } else {
                        console.error(e.message);
                    }
                } else {
                    console.error(e);
                }
            }
        }, searchTypingDelay),
        [controller]
    );

    const displayCss = showDropdownIndicator ? "inherit" : "none";
    const chakraStyles = {
        dropdownIndicator: (provided: object) => ({
            ...provided,
            display: displayCss
        })
    };

    //Typescript doesn't recognize Type SearchOption but rather SingleValue<SearchGroupOption>
    const onInputChange = (value: unknown, actionMeta: ActionMeta<unknown>) => {
        if (value && actionMeta.action === "select-option") {
            onSelect({ action: "select", value: value as SearchOption });
        } else if (actionMeta.action === "clear") {
            onClear({
                action: "select",
                value: actionMeta.removedValues as unknown as SearchOption
            });
        } else {
            console.debug("unknown Actiontype");
        }
    };

    return (
        <Box {...containerProps}>
            <FormControl alignItems="center">
                <AsyncSelect
                    isClearable={true}
                    placeholder={placeholder}
                    closeMenuOnSelect={closeMenuOnSelect}
                    loadOptions={debouncedLoadOptions}
                    components={{
                        Option: HighlightOption,
                        NoOptionsMessage: NoOptionsMessage,
                        Menu: MenuComp,
                        LoadingMessage: LoadingMessage
                    }}
                    chakraStyles={chakraStyles}
                    onChange={onInputChange}
                />
            </FormControl>
        </Box>
    );
};

// TODO: This should accept all functions as an utility function, but this is banned by eslint.
// TODO: It would be better when the delay would not trigger the loading animation
// eslint-disable-next-line @typescript-eslint/ban-types
function debounce<T extends Function>(delayedFunction: T, delay = 250) {
    let timeout: NodeJS.Timeout | string | number | undefined;

    return (...args: unknown[]) => {
        timeout && clearTimeout(timeout);
        timeout = setTimeout(() => {
            delayedFunction(...args);
        }, delay);
    };
}

function mapSuggestions(suggestions: Suggestion[][], sources: DataSource[]) {
    const options = sources.map((source, index) => ({
        label: source.label,
        options: suggestions[index]?.map((item) => ({ value: item.text, label: item.text })) || []
    }));
    return options;
}
function useController(sources: DataSource[], map: MapModel | undefined) {
    const [controller, setController] = useState<SearchController | undefined>(undefined);
    useEffect(() => {
        if (!map) {
            return;
        }
        const controller = new SearchController({ sources });
        setController(controller);
        return () => {
            setController(undefined);
        };
    }, [map, sources]);

    return controller;
}

export const MenuComp = (props: MenuProps<SearchGroupOption>) => {
    const hasInput = props.selectProps.inputValue.length > 0;
    let clazz = "";
    if (!hasInput) {
        clazz = "search-invisible";
    }
    return (
        <components.Menu {...props} className={clazz}>
            {props.children}
        </components.Menu>
    );
};

export const NoOptionsMessage = (props: NoticeProps<SearchGroupOption>) => {
    const intl = useIntl();
    // TODO: Make it configurable?
    const noMessageText = intl.formatMessage({ id: "noOptionsText" });

    return (
        <components.NoOptionsMessage {...props}>
            <chakra.span className="search-no-match">{noMessageText}</chakra.span>
        </components.NoOptionsMessage>
    );
};

export const LoadingMessage = (props: NoticeProps<SearchGroupOption>) => {
    const intl = useIntl();
    // TODO: Make it configurable?
    const loadingText = intl.formatMessage({ id: "loadingText" });

    return (
        <components.LoadingMessage {...props}>
            <chakra.span className="search-loading-text">{loadingText}</chakra.span>
        </components.LoadingMessage>
    );
};
