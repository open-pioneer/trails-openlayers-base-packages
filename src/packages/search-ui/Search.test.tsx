// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0

import { expect, it } from "vitest";
import { createServiceOptions, setupMap } from "@open-pioneer/map-test-utils";
import { PackageContextProvider } from "@open-pioneer/test-utils/react";
import { render, screen, waitFor } from "@testing-library/react";
import { Search, SearchEvent } from "./Search";
import { FakeCitySource, FakeRiverSource, FakeStreetSource } from "./testSources";
import userEvent from "@testing-library/user-event";

it("should successfully create a measurement component", async () => {
    await createSearch();
    // search is mounted
    const { searchDiv } = await waitForSearch();
    expect(searchDiv).toMatchSnapshot();
});

it("should successfully type into search", async () => {
    const user = userEvent.setup();
    await createSearch();
    // search is mounted
    const { searchInput } = await waitForInput();
    await user.type(searchInput, "Dortmund");
    expect(searchInput).toHaveValue("Dortmund");
});

it.skip("should successfully show a search suggestion", async () => {
    const user = userEvent.setup();

    await createSearch();

    // search is mounted
    const { searchInput } = await waitForInput();
    await user.type(searchInput, "Dortmund");

    // FIX ME: Do not use timeout here!! Too long and buggy
    // Always ensure that test code happens in the async unit test function!
    setTimeout(async () => {
        const { suggestion } = await waitForSuggestion();
        expect(suggestion); //.toHaveValue("Dortmund");
    }, 1000);
});

//TODO: remove fails
it.fails("should successfully show a suggestion select", async () => {
    const user = userEvent.setup();

    const selectHandler = vi.fn();

    await createSearch(selectHandler);
    // search is mounted
    const { searchInput } = await waitForInput();
    await user.type(searchInput, "Dortmund");
    
    const { suggestion } = await waitForSuggestion();
    await userEvent.click(suggestion);
    //Todo Test not Work
    expect(selectHandler).toBeCalledTimes(1);
});

//TODO: remove fails
it.fails("should successfully clear a suggestion select", async () => {
    const user = userEvent.setup();

    const selectHandler = vi.fn();
    const clearHandler = vi.fn();

    await createSearch(selectHandler, clearHandler);
    // search is mounted
    const { searchInput } = await waitForInput();
    await user.type(searchInput, "Dortmund");
    
    const { suggestion } = await waitForSuggestion();
    await userEvent.click(suggestion);
    const { clearButton } = await waitForClearButton();
    await userEvent.click(clearButton);
    //Todo Test not Work
    expect(clearHandler).toBeCalledTimes(1);
});

async function createSearch(
    selectHandler?: (event: SearchEvent) => void,
    clearHandler?: (event: SearchEvent) => void
) {
    const { mapId, registry } = await setupMap();
    await registry.expectMapModel(mapId);
    const injectedServices = createServiceOptions({ registry });
    const sources = [new FakeCitySource(), new FakeRiverSource(), new FakeStreetSource()];
    const selectHandlerFunction = selectHandler? selectHandler:(_event: SearchEvent) => {};
    const clearHandlerFunction = clearHandler? clearHandler: (_event: SearchEvent) => {};
    render(
        <PackageContextProvider services={injectedServices}>
            <Search
                data-testid="search"
                mapId={mapId}
                sources={sources}
                onSelect={selectHandlerFunction}
                onClear={clearHandlerFunction}
            ></Search>
        </PackageContextProvider>
    );
}

async function waitForSearch() {
    const searchDiv =
        await screen.findByTestId<HTMLDivElement>("search");

    return { searchDiv };
}

async function waitForInput() {
    const { searchDiv } = await waitForSearch();
    const searchInput = searchDiv.getElementsByTagName("input")[0];
    if (!searchInput) {
        throw new Error("input not rendered");
    }
    return { searchInput };
}

async function waitForSuggestion() {
    const { suggestion } = await waitFor(async () => {
        const { searchDiv } = await waitForSearch();
        const suggestion = searchDiv.getElementsByClassName("search-highlighted-match")[0];

        if (!suggestion) {
            throw new Error("Suggestion not found");
        }
        return { suggestion };
    });
    return { suggestion }; 
}

async function waitForClearButton() {
    const { clearButton } = await waitFor(async () => {
        const clearButton: Element | null = (
            await screen.findByTestId<HTMLDivElement>("search")
        ).querySelector('[role = "button"]');

        if (!clearButton) {
            throw new Error("Clearbutton not found");
        }
        return { clearButton };
    });
    return { clearButton };
}
