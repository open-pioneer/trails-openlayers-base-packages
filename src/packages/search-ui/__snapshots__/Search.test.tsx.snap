// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should successfully create a measurement component 1`] = `
<div
  class="search css-0"
  data-testid="search"
  data-theme="light"
>
  <div
    class="chakra-form-control css-hnw0d4"
    data-theme="light"
    role="group"
  >
    <div
      class=" css-79elbk"
      data-theme="light"
    >
      <span
        class="css-1f43avz-a11yText-A11yText"
        id="react-select-2-live-region"
      />
      <span
        aria-atomic="false"
        aria-live="polite"
        aria-relevant="additions text"
        class="css-1f43avz-a11yText-A11yText"
      />
      <div
        class=" css-i2418r"
        data-theme="light"
      >
        <div
          class=" css-j93siq"
          data-theme="light"
        >
          <div
            class=" css-hji3d9"
            data-theme="light"
            id="react-select-2-placeholder"
          >
            Select...
          </div>
          <div
            class=" css-18euh9p"
            data-theme="light"
            data-value=""
          >
            <input
              aria-autocomplete="list"
              aria-describedby="react-select-2-placeholder"
              aria-expanded="false"
              aria-haspopup="true"
              aria-required="false"
              autocapitalize="none"
              autocomplete="off"
              autocorrect="off"
              class=" css-10wwmqn"
              data-theme="light"
              id="field-:r0:"
              role="combobox"
              spellcheck="false"
              tabindex="0"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class=" css-hfbj6y"
          data-theme="light"
        >
          <hr
            aria-orientation="vertical"
            class="chakra-divider css-1i6c5ox"
            data-theme="light"
          />
          <div
            aria-hidden="true"
            class=" css-80pt9o"
            data-theme="light"
          >
            <svg
              aria-hidden="true"
              class="chakra-icon css-onkibi"
              data-theme="light"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
